brave-tests/src/test/java/brave/propagation/ThreadLocalCurrentTraceContextClassLoaderTest.java

```diff
@@ -43,9 +43,24 @@
     }
   }
 
+  @Test public void leakedNullScope() {
+    assertRunIsUnloadable(LeakedNullScope.class, getClass().getClassLoader());
+  }
+
+  static class LeakedNullScope implements Runnable {
+    @Override public void run() {
+      CurrentTraceContext current = ThreadLocalCurrentTraceContext.newBuilder().build();
+      current.newScope(null);
+    }
+  }
+
   /**
-   * TODO: While it is an instrumentation bug to leak a scope, we should be tolerant, for example
-   * considering weak references or similar.
+   * TODO: While it is an instrumentation bug to leak a scope, we should be tolerant.
+   *
+   * <p>The current design problem is we don't know a reference type we can use that clears when
+   * the classloader is unloaded, regardless of GC. For example, having {@link Scope} extend {@link
+   * java.lang.ref.WeakReference} to hold the value to revert. This would only help if GC happened
```

Comment:
Doesn't block this PR at all, just to record my thought and maybe @felixbarny has some ideas too.

I haven't worked with unloading scenarios much but wonder whether this is so odd - since most things get cleaned up when GC happens, it seems natural to me for unloaded classes to also wait until GC to be cleaned up.